app/models/role.rb

```diff
@@ -67,11 +67,29 @@ def by_login login, account:
   end
   
   def password= password
-    self.credentials ||= Credentials.new(role: self)
-    self.credentials.password = password
-    self.credentials.save(raise_on_save_failure: true)
+    modify_credentials do |credentials| 
+      credentials.password = password
+    end
   end
-  
+
+  def valid_origin?(ip_addr)
+    ip = IPAddr.new(ip_addr)
+    restricted_to.blank? || restricted_to.any? do |cidr|
+      cidr.include?(ip)
+    end
+  end
+
+  def restricted_to
+    credentials = self.credentials ||= Credentials.new(role: self)
```

Comment:
I feel like the extra local variable here and in `#modify_credentials` takes away from the readability of the code. Maybe just use `self.credentials` instead?